### PR TITLE
USD Crate format Documentation

### PR DESCRIPTION
**Pull request by dgovil**
### Description of Change(s)

This PR adds an initial pass of the Crate format specification. Where possible it tries to describe things at a byte or bit level.
This is a fairly large undertaking to reverse out the code into natural language, so requires extra scrutiny.

This adds a few features:

* Custom sphinx directive to read marked comments within the C++ source code. This allows the majority of the documentation to live beside the implementation. It also allows for ordering to change as needed which is handy. 
This enables Alex's request to have something akin to literate programming approach within the code.
* Custom sphinx directive to generate byte layouts, inspired by the [PKZip documentation.](https://users.cs.jmu.edu/buchhofp/forensics/formats/pkzip.html)

Sphinx was chosen to include this documentation because:

* The other specifications are in Sphinx already
* Doxygen has limited support for custom documentation extensions

The document includes some warnings for now. I think it would be best to leave the warnings in place till we are certain that the descriptions are accurate. Beyond that, it might be good to take a leaf out of other project specifications books where the official implementation is the gold truth and warnings explicitly state that.

Please note that this PR also includes https://github.com/PixarAnimationStudios/USD/pull/2084 because it's easier on my eyes than the default RTD theme, and provides better visual sectioning differences as well, which are necessary for how deep the nesting levels get.

A PDF preview is attached but the PDF formatting gets messed up so it's still recommended to sphinx-build the docs and view it that way. 
[USDC File Format Specification — Universal Scene Description 22.11 documentation.pdf](https://github.com/PixarAnimationStudios/USD/files/9984656/USDC.File.Format.Specification.Universal.Scene.Description.22.11.documentation.pdf)

- [X ] I have verified that all unit tests pass with the proposed changes

- [X ] I have submitted a signed Contributor License Agreement

### CODE REVIEW

File: pxr/base/tf/fastCompression.cpp [hunk 1]
```diff
@@ -101,6 +101,24 @@ TfFastCompression::CompressToBuffer(
     return compressed - origCompressed;
 }    
 
+/* usddoc lz4
+ * The first byte of any buffer passed to be decrypted stores the number of chunks used.
```
Comment (meshula) [inline pxr/base/tf/fastCompression.cpp]:
```suggestion
 * The first byte of any buffer passed to be decompressed stores the number of chunks used.
```

File: pxr/usd/usd/crateFile.cpp [hunk 2]
```diff
@@ -1861,6 +2014,18 @@ _ReadPossiblyCompressedArray(
 
 struct _CompressedIntsReader
 {
+    /* usddoc compressed_int
+     *  Compressed integers are stored as a contiguous , homogenous array of
```
Comment (meshula) [inline pxr/usd/usd/crateFile.cpp]:
```suggestion
     *  Compressed integers are stored as a contiguous, homogenous array of
```

File: pxr/usd/usd/crateFile.cpp [hunk 3]
```diff
@@ -3235,6 +3448,16 @@ CrateFile::_ReadStructuralSections(Reader reader, int64_t fileSize)
     if (m.IsClean()) _ReadSpecs(reader);
 }
 
+/* usddoc bootstrap
+ * Every USD file starts with a Bootstrap that tells us information to parse the file.
+ *
+ * Files must start with an identifier (PXR-USDC), the version number and the offset to the Table of Contents.
+ * There is additional reserved space at the end of the bootstrap for future use.
+ *
+ * Version numbers are stored as Major , Minor and Patch followed by unused bytes.
```
Comment (meshula) [inline pxr/usd/usd/crateFile.cpp]:
```suggestion
 * Version numbers are stored as Major, Minor and Patch followed by unused bytes.
```
the table shows 8  bytes for version, how are major, minor, and patch, packed into those 8 bytes?

File: pxr/usd/usd/crateFile.h [hunk 4]
```diff
@@ -79,13 +79,36 @@ template <class T> struct _IsBitwiseReadWrite;
...
-// Value in file representation.  Consists of a 2 bytes of type information
-// (type enum value, array bit, and inlined-value bit) and 6 bytes of data.
-// If possible, we attempt to store certain values directly in the local
-// data, such as ints, floats, enums, and special-case values of other types
-// (zero vectors, identity matrices, etc).  For values that aren't stored
-// inline, the 6 data bytes are the offset from the start of the file to the
-// value's location.
+/* usddoc value_rep
+ * Crate stores Value Representations as data blobs that are read on demand.
+ * This allows large USD scenes to be read incredibly quickly by deferring all
```
Comment (meshula) [inline pxr/usd/usd/crateFile.h]:
```suggestion
 * This allows large USD scenes to be read as quickly as possible by deferring all
```
avoiding subjective language

File: pxr/usd/usd/crateFile.h [hunk 5]
```diff
@@ -79,13 +79,36 @@ template <class T> struct _IsBitwiseReadWrite;
...
+ * data reads to the latest point possible.
+ *
+ * Value reps are stored as an unsigned 64 bit integer.
```
Comment (meshula) [inline pxr/usd/usd/crateFile.h]:
```suggestion
 * Value representations are stored as an unsigned 64 bit integer.
```
prefer full words to aid full text indexing and search

File: pxr/usd/usd/crateFile.cpp [hunk 6]
```diff
@@ -1936,6 +2107,21 @@ _ReadPossiblyCompressedArray(
...
+ * Compressed float arrays start with an unsigned 64 bit integer to represent their element count.
+ *
+ * Arrays that are 16 bytes or smaller are not compressed.
+ *
+ * Following this, a char is used to represent the array encoding scheme:
+ *
+ * * **i** for integer compression (follow the compressed integer array logic)
+ * * **t** for lookup tables (described below)
+ *
+ * The size of the lookup table (LUT) is a 32bit unsigned integer.
```
Comment (meshula) [inline pxr/usd/usd/crateFile.cpp]:
```suggestion
 * The size of the lookup table (LUT) is a 32 bit unsigned integer.
```

File: pxr/usd/usd/crateFile.cpp [hunk 7]
```diff
@@ -1179,16 +1213,58 @@ class CrateFile::_Reader : public _ReaderBase
...
+     * stored as contiguous key:value pairs.
+     */
     SdfVariantSelectionMap Read(SdfVariantSelectionMap *) {
         return ReadMap<SdfVariantSelectionMap>();
     }
+    /* usddoc layer_offset
+     * A layer offset represents a time offset and scale between layers.
+     * It is represented by two doubles.
+     *
+     * The first represents the time offset to be used , and the second the scale.
```
Comment (meshula) [inline pxr/usd/usd/crateFile.cpp]:
```suggestion
     * The first represents the time offset to be used, and the second the scale.
```

File: pxr/usd/usd/crateFile.cpp [hunk 8]
```diff
@@ -430,6 +430,20 @@ struct _PathItemHeader {
...
 struct _IsBitwiseReadWrite<_PathItemHeader> : std::true_type {};
 
+/* usddoc listop_header
+ * ListOps have an 8-byte bitmasked header that determine what type of operation
+ * is being run.
+ *
+ * In order, from least to most significant bit:
+ *
+ *  * **IsExplicit** : Removes all items and changes the list to be explicit
+ *  * **HasExcplicitItems** : Fills the list with the items included in this ListOp
```
Comment (meshula) [inline pxr/usd/usd/crateFile.cpp]:
```suggestion
 *  * **HasExplicitItems** : Fills the list with the items included in this ListOp
```

**Review by meshula — COMMENTED** [review]
Exciting! This review has my notes

not sure where the string in the pdf, "Integers are stored with the least significant bit", comes from, but should read "least  significant byte".

Can the "Certain types" "Internal to Pixar" be flagged in some way in the table that follows?

"For multidimensional type like Matrix2D".... The explanation and diagram that follows suggest the diagonal only is stored as three entries? Very confusing as one expects to find nine values, and one would want to know something like, "matrix rows are stored consecutively in memory such that they can be addressed  by the C syntax,  matrix[row][column]" Some note about whether these are mathematically column major matrices stored in C row major order is probably helpful.

In Base Types, Bool should note if it is stored as an 8-bit value where the values of 0 or 1 are valid (needs double checking, I'm not sure). Also, some notes on bool array ~ are they bytes, or bit packed (I haven't checked)

**Comment by dgovil** [comment]
Thanks, addressed some of your notes.

With regards to some of the remaining notes:

* That specific Matrix diagram  you mentioned is for inlined types, which only seem to store the diagonal values. I do mention the row major nature of Matrix values in the Matrix section lower down, but I added an additional note up here about it for clarity since Inlined types show up first. Since there'd only be 6 bytes left, only a matrix2D would fit fully inlined and there's no specialization for it in the code.

* For binary arrays, I'm not sure and I can't seem to find the exact implementation reader in the code. Maybe @gitamohr can shed some light? I believe it's not a bit packed vector since I don't see a specialization anywhere.

**Comment by dgovil** [comment]
Also @asluk had asked for a zip of the sphinx doc, so here's the zip. You'll want to open `specs_usdc.html`
[usddocs.zip](https://github.com/PixarAnimationStudios/USD/files/9985693/usddocs.zip)

**Comment by meshula** [comment]
Thanks for the update.

@gitamohr The specific question was the encoding of bools, and arrays of bools.

**Comment by syoyo** [comment]
> @gitamohr The specific question was the encoding of bools, and arrays of bools.

Array of bools `bool[]` is encoded as array of bytes(8bit), since bool array is represented as `VtArray<bool>`  and  `VtArray<bool>` uses `sizeof(bool) = 1` as an element size to allocate memory: https://github.com/PixarAnimationStudios/USD/blob/0c7b9a95f155c221ff7df9270a39a52e3b23af8b/pxr/base/vt/array.h#L854

For scalar `bool` type, datasize could be arbitrary, since it is represented as inlined value and fits to ValueRep's 6bytes payload.

**Comment by dgovil** [comment]
Thanks @syoyo , I updated the section on Bool with that note, and also thanks because I used both USD and your TinyUSDZ project as a reference when writing these docs. Having the two was very helpful.

**Comment by syoyo** [comment]
@dgovil 

>  Having the two was very helpful.

🤩👍

**Comment by sunyab** [comment]
Filed as internal issue #USD-7771

File: docs/spec_usdc.rst [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
++--------------------+------------+----------------------------------------------------+
+| customLayerData    | dictionary | Custom user specified data                         |
++--------------------+------------+----------------------------------------------------+
+| defaultPrim        | token      | The prim to use when this layer is referenced      |
++--------------------+------------+----------------------------------------------------+
+| documentation      | string     | Top level documentation for this layer             |
++--------------------+------------+----------------------------------------------------+
+| metersPerUnit      | double     | The scale unit for this layer                      |
++--------------------+------------+----------------------------------------------------+
+| primChildren       | token[]    | A list of top level children to limit the layer to |
```
Comment (syoyo) [inline docs/spec_usdc.rst:233]: `primChildren` and `properties` in Layer and Prim meta uses TokenVector(41), not array of `token`

```
name = "primChildren", value = ty: 41, isArray: 0, isInlined: 0, isCompressed: 0 
```

**Comment by syoyo** [comment]
@dgovil Added comment on type for `primChildren` and `properties` https://github.com/PixarAnimationStudios/USD/pull/2086/files#r1027088026

TokenVector and `token[]`  have different meaning in pxrUSD implementation.
For example, you cannot use TokenVector to encode `token[]` attribute. When I encode a Prim attribute with `token[]` type as a TokenVector, pxrUSD apprently loads it finely but prints corrupted info in ASCII like this:

```
uniform token[] jointNames = <'vector<TfToken, allocator<TfToken> >' @ 0x55a0ab36c450>
```

**Comment by spiffmon** [comment]
Hi @dgovil - have been meaning to provide some high level feedback that @gitamohr and I agreed on.  We think it would be overall clarifying and more economical to **first** provide a specification for the Sdf Data Model, which defines all terms, concepts, types, etc.  Then, each of the text-format and crate specifications can reference and build off of that shared data model specification.  Having the data model pulled out on its own would also be a boon to anyone trying to adapt a foreign data format to USD through an `SdfFileFormat` plugin.

What do you think?

**Comment by dgovil** [comment]
I think that makes sense. Would just a Sphinx rst page make sense for something like that? Since it wouldn't be tied to any code etc

**Comment by spiffmon** [comment]
Yes, absolutely - though it might someday become part of something larger
hosted/interred elsewhere, a new page under our current "Specifications"
section makes sense to me for now?

--spiff

**Pull request closed without merging**